### PR TITLE
Switching back to using tags in email alert payload

### PR DESCRIPTION
**Pull request by christophwong**
[TRELLO](https://trello.com/c/ploRXLJV)
As of now email-alert-api does not support a way without using "tags" to get a
multi-document subscriber list (subscribed to all /cma-cases changes as
opposed to one specific document as seen with travel advise publisher's
usage) from content-ids/links.

Leaving links intact because a.) it doesn't affect the payload in initial testing, b.) tags are being deprecated and we hope to use just links in the future.

I have tested this change in payload format by using 

```
curl -X POST -H "Content-Type: application/json" -d <actual payload> https://email-alert-api.integration.publishing.service.gov.uk/notifications
```

The payload where marked <actual payload> in the above example is as follow with the HTML replaced with <shortened> (but for context the actual html string was used). Focus is on the `"tags"` attribute

```
{"subject": "test with b", "body": "<shortened>", "links": {"topic": ["e033acc7-9034-4f01-90f4-ca6a49952194"]}, "tags": {"format": "cma_case", "opened_date": "2016-04-10", "case_type": "ca98-and-civil-cartels", "case_state": "open", "market_sector":["something"]}, "document_type": "cma_case", "header": "<shortened>", "footer": "<shortened>"}
```

In the Tech Debt session of email-alert-api govuk docs, regarding the new links architechure, one of the future goals is "Resolving how specialist document alerts would work alongside such an architecture"
Reference: https://gov-uk.atlassian.net/wiki/display/TECH/Email+Alert+API

### CODE REVIEW

**Comment by mathildathompson** [comment]
👍 Looks good to me. Think it would be useful for someone (maybe @elliotcm ) with a bit more context of email-alert-api to double check.

File: app/presenters/email_alert_presenter.rb [hunk 1]
```diff
@@ -24,6 +25,19 @@ def body
     end
   end
 
+  def tags
+    { format: document.publishing_api_document_type }.deep_merge(metadata)
+  end
+
+  def metadata
```
Comment (benilovj) [inline app/presenters/email_alert_presenter.rb]: I find this method a bit cryptic. Do the fields have to be dynamic here, or are only a few used (eg `format`, `case_type`)?

**Comment by elliotcm** [comment]
I'd recommend removing the links because while I think it'll be safe, we work under the assumption that it'll be one or the other when triggering notifications.  Having both in subscriptions is OK.

File: app/presenters/email_alert_presenter.rb (hunk 1 above)
Comment (christophwong) [inline app/presenters/email_alert_presenter.rb]: The reason I went with a dynamic approach was because for each document type, the tags (with the exception of format) are different. So I copied this pattern from document presenter.
`case_type` is only used for cma_cases, and for each document type, a different combinations of tags are used.

Comment (benilovj) [inline app/presenters/email_alert_presenter.rb]: IMHO, if it has to be dynamic, then this is fine, but if it could be static, an explicit list of fields would read better. @elliotcm any thoughts? I don't really have enough context here.

**Comment by christophwong** [comment]
@elliotcm links has been removed.

File: app/presenters/email_alert_presenter.rb (hunk 1 above)
Comment (elliotcm) [inline app/presenters/email_alert_presenter.rb]: You're both right!  Pulling them from the objects looks to be the way it's got to be done, but it is quite a hard to parse method.  @christophwong could you spend a short time seeing if it can be made clearer?

**Comment by elliotcm** [comment]
@christophwong Thanks.

File: app/presenters/email_alert_presenter.rb (hunk 1 above)
Comment (christophwong) [inline app/presenters/email_alert_presenter.rb]: Gladly, if we find a solution to this, we can also clean up the SearchPresenter and DocumentPresenter too.

I think I will use this [method](https://github.com/alphagov/specialist-publisher-rebuild/blob/master/app/models/document.rb#L92) in my first step, (although it still has the same problem of lacking clarity, I still believe "metadata" belongs in the document model rather than the presenter.)

File: app/presenters/email_alert_presenter.rb [hunk 2]
```diff
@@ -24,6 +24,12 @@ def body
     end
   end
 
+  def tags
+    {
+      format: document.publishing_api_document_type
+    }.deep_merge(document.format_specific_metadata.reject { |_k, v| v.blank? })
```
Comment (christophwong) [inline app/presenters/email_alert_presenter.rb:30]: @benilovj @elliotcm this is the refactoring (it has caused the previous discussion to be hidden) but the problem of clarity is still here.
However, I am leaning towards dynamic approach towards the metadata, as writing all the fields out for each specific format seemed messy. 

Also, in this system the attributes that are grouped into "format_specific_metadata" seems semantically accurate to me. I believe our app only needs to know them as a group of metadata, and there isn't specific information or behavioural difference between the specific fields inside this metadata group.

Comment (christophwong) [inline app/presenters/email_alert_presenter.rb:30]: similar pattern:
- [Document Presenter](https://github.com/alphagov/specialist-publisher-rebuild/blob/master/app/presenters/document_presenter.rb#L51-L63)
- [SearchPresenter for rummager](https://github.com/alphagov/specialist-publisher-rebuild/blob/master/app/presenters/search_presenter.rb#L16)

Comment (elliotcm) [inline app/presenters/email_alert_presenter.rb:30]: Seems OK to me.

Comment (christophwong) [inline app/presenters/email_alert_presenter.rb:30]: ok to merge?

Comment (elliotcm) [inline app/presenters/email_alert_presenter.rb:30]: @benilovj?
